packages/api/amplify_api_dart/lib/src/api_plugin_impl.dart

```diff
@@ -0,0 +1,394 @@
+import 'package:amplify_api_dart/src/graphql/providers/oidc_function_api_auth_provider.dart';
+import 'package:amplify_api_dart/src/graphql/web_socket/blocs/web_socket_bloc.dart';
+import 'package:amplify_api_dart/src/graphql/web_socket/services/web_socket_service.dart';
+import 'package:amplify_api_dart/src/graphql/web_socket/state/web_socket_state.dart';
+import 'package:amplify_api_dart/src/graphql/web_socket/types/connectivity_platform.dart';
+import 'package:amplify_api_dart/src/util/amplify_api_config.dart';
+import 'package:amplify_api_dart/src/util/amplify_authorization_rest_client.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:meta/meta.dart';
+
+/// {@template amplify_api_dart.amplify_api_dart}
+/// The AWS implementation of the Amplify API category in Dart.
+/// {@endtemplate}
+class AmplifyAPIDart extends APIPluginInterface with AWSDebuggable {
+  /// {@macro amplify_api_dart.amplify_api_dart}
+  AmplifyAPIDart({
+    List<APIAuthProvider> authProviders = const [],
+    ConnectivityPlatform connectivity = const ConnectivityPlatform(),
+    AWSHttpClient? baseHttpClient,
+    this.modelProvider,
+    this.subscriptionOptions,
+  })  : _baseHttpClient = baseHttpClient,
+        _connectivity = connectivity {
+    authProviders.forEach(registerAuthProvider);
+    Amplify.Hub.addChannel(HubChannel.Api, _hubEventController.stream);
+  }
+
+  late final AWSApiPluginConfig _apiConfig;
+  final AWSHttpClient? _baseHttpClient;
+  late final AmplifyAuthProviderRepository _authProviderRepo;
+
+  /// Creates a stream representing network connectivity at the hardware level.
+  final ConnectivityPlatform _connectivity;
+
+  /// A map of the keys from the Amplify API config with auth modes to HTTP clients
+  /// to use for requests to that endpoint/auth mode. e.g. { "myEndpoint.AWS_IAM": AWSHttpClient}
+  final Map<String, AWSHttpClient> _clientPool = {};
+
+  /// A map of the keys from the Amplify API config websocket connections to use
+  /// for that endpoint.
+  final Map<String, WebSocketBloc> _webSocketBlocPool = {};
+
+  /// The registered [APIAuthProvider] instances.
+  final Map<APIAuthorizationType, APIAuthProvider> _authProviders = {};
+
+  final StreamController<ApiHubEvent> _hubEventController =
+      StreamController<ApiHubEvent>.broadcast();
+
+  /// Subscription options
+  final GraphQLSubscriptionOptions? subscriptionOptions;
+
+  @override
+  Future<void> reset() async {
+    for (final bloc in _webSocketBlocPool.values) {
+      bloc.add(const ShutdownEvent());
+    }
+
+    await Future.wait(
+      _webSocketBlocPool.values.map((bloc) => bloc.done.future),
+    );
+
+    await _hubEventController.close();
+  }
+
+  @override
+  Future<void> configure({
+    AmplifyConfig? config,
+    required AmplifyAuthProviderRepository authProviderRepo,
+  }) async {
+    final apiConfig = config?.api?.awsPlugin;
+    if (apiConfig == null) {
+      throw const ApiException(
```

Comment:
nit: should thrown `ConfigurationException` align with [other plugins](https://github.com/aws-amplify/amplify-flutter/blob/next/packages/auth/amplify_auth_cognito_dart/lib/src/auth_plugin_impl.dart#L275)?

(storage refactoring for this is coming.)